src/H5VLint.c

```diff
@@ -759,189 +782,159 @@ H5VL_register_using_existing_id(H5I_type_t type, void *object, H5VL_t *vol_conne
-    H5VL_t       *connector    = NULL;  /* New VOL connector struct */
-    bool          conn_id_incr = false; /* Whether the VOL connector ID has been incremented */
-    H5VL_t       *ret_value    = NULL;  /* Return value */
+    H5VL_object_t *ret_value = NULL; /* Return value */
 
     FUNC_ENTER_NOAPI(NULL)
 
-    /* Get the VOL class object from the connector's ID */
-    if (NULL == (cls = (H5VL_class_t *)H5I_object_verify(connector_id, H5I_VOL)))
-        HGOTO_ERROR(H5E_VOL, H5E_BADTYPE, NULL, "not a VOL connector ID");
+    /* Check arguments */
+    assert(object);
+    assert(vol_connector);
 
-    /* Setup VOL info struct */
-    if (NULL == (connector = H5FL_CALLOC(H5VL_t)))
-        HGOTO_ERROR(H5E_VOL, H5E_CANTALLOC, NULL, "can't allocate VOL connector struct");
-    connector->cls = cls;
-    connector->id  = connector_id;
-    if (H5I_inc_ref(connector->id, false) < 0)
-        HGOTO_ERROR(H5E_VOL, H5E_CANTINC, NULL, "unable to increment ref count on VOL connector");
-    conn_id_incr = true;
+    /* Set up VOL object for the passed-in data */
+    /* (Does not wrap object, since it's from a VOL callback) */
+    if (NULL == (ret_value = H5FL_CALLOC(H5VL_object_t)))
+        HGOTO_ERROR(H5E_VOL, H5E_CANTALLOC, NULL, "can't allocate memory for VOL object");
+    ret_value->connector = vol_connector;
+    ret_value->data      = object;
+    ret_value->rc        = 1;
 
-    /* Set return value */
-    ret_value = connector;
+    /* Bump the reference count on the VOL connector */
+    H5VL_conn_inc_rc(vol_connector);
 
 done:
-    /* Clean up on error */
-    if (NULL == ret_value) {
-        /* Decrement VOL connector ID ref count on error */
-        if (conn_id_incr && H5I_dec_ref(connector_id) < 0)
-            HDONE_ERROR(H5E_VOL, H5E_CANTDEC, NULL, "unable to decrement ref count on VOL connector");
-
-        /* Free VOL connector struct */
-        if (NULL != connector)
-            connector = H5FL_FREE(H5VL_t, connector);
-    } /* end if */
-
     FUNC_LEAVE_NOAPI(ret_value)
-} /* end H5VL_new_connector() */
+} /* end H5VL_create_object() */
 
 /*-------------------------------------------------------------------------
- * Function:	H5VL_register_using_vol_id
+ * Function:	H5VL__conn_create
  *
- * Purpose:     Utility function to create a user ID for an object created
- *              or opened through the VOL. Uses the VOL connector's ID to
- *              get the connector information instead of it being passed in.
+ * Purpose:     Utility function to create a connector around a class
  *
- * Return:      Success:    A valid HDF5 ID
- *              Failure:    H5I_INVALID_HID
+ * Return:      Success:    Pointer to a new connector object
+ *              Failure:    NULL
  *
  *-------------------------------------------------------------------------
  */
-hid_t
-H5VL_register_using_vol_id(H5I_type_t type, void *obj, hid_t connector_id, bool app_ref)
```

Comment:
H5VL_register_using_vol_id is only used in the testing code now, so I moved it to H5VLtest.c